src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.ImmutableTypeParameterArgumentAnalysis.cs

```diff
@@ -0,0 +1,642 @@
+							immutabilityContext,
+							parameter.Type,
+							() => {
+								return syntax
+									.ParameterList
+									.Parameters[ i ]
+									.Type;
+							}
+						);
+					}
+				},
+				SyntaxKind.LocalFunctionStatement
+			);
+
+			// Type arguments of method invocations
+			// Type arguments of containing types for statically invoked methods
+			context.RegisterOperationAction(
+				ctx => {
+					var operation = (IInvocationOperation)ctx.Operation;
+
+					SyntaxNodeOrToken getSyntax() => operation.Syntax switch {
+						InvocationExpressionSyntax invocation => invocation.Expression,
+						_ => operation.Syntax
+					};
+
+					AnalyzeMethodUsage(
+						ctx,
+						annotationsContext,
+						immutabilityContext,
+						operation.TargetMethod,
+						getSyntax
+					);
+				},
+				OperationKind.Invocation
+			);
+
+			// Type arguments of method references
+			// Type arguments of containing types for statically referenced methods
+			context.RegisterOperationAction(
+				ctx => {
+					var operation = (IMethodReferenceOperation)ctx.Operation;
+
+					AnalyzeMethodUsage(
+						ctx,
+						annotationsContext,
+						immutabilityContext,
+						operation.Method,
+						() => operation.Syntax
+					);
+				},
+				OperationKind.MethodReference
+			);
+
+			// Type arguments of field types
+			context.RegisterSymbolAction(
+				ctx => {
+					var symbol = (IFieldSymbol)ctx.Symbol;
+
+					SyntaxNodeOrToken getSyntax() {
+						SyntaxNode syntaxNode = symbol.DeclaringSyntaxReferences[ 0 ].GetSyntax( ctx.CancellationToken );
+
+						return syntaxNode switch {
+							FieldDeclarationSyntax fieldDeclaration => fieldDeclaration.Declaration.Type,
+							VariableDeclaratorSyntax variableDeclarator => ( (VariableDeclarationSyntax)variableDeclarator.Parent! ).Type,
+							_ => syntaxNode,
+						};
+					}
+
+					AnalyzeTypeRecursive(
+						ctx.ReportDiagnostic,
+						annotationsContext,
+						immutabilityContext,
+						symbol.Type,
+						getSyntax
+					);
+				},
+				SymbolKind.Field
+			);
+
+			// Type arguments of containing types of statically referenced fields
+			context.RegisterOperationAction(
+				ctx => {
+					var operation = (IFieldReferenceOperation)ctx.Operation;
+
+					AnalyzeMaybeStaticMemberAccess(
+						ctx,
+						annotationsContext,
+						immutabilityContext,
+						operation.Field,
+						() => operation.Syntax
+					);
+				},
+				OperationKind.FieldReference
+			);
+
+			// Type arguments of property types
+			context.RegisterSymbolAction(
+				ctx => {
+					var symbol = (IPropertySymbol)ctx.Symbol;
+
+					if( symbol.Type is not INamedTypeSymbol type ) {
+						return;
+					}
+
+					SyntaxNodeOrToken getSyntax() {
+						SyntaxNode syntaxNode = symbol.DeclaringSyntaxReferences[ 0 ].GetSyntax( ctx.CancellationToken );
+
+						return syntaxNode switch {
+							IndexerDeclarationSyntax indexer => indexer.Type,
+							ParameterSyntax parameter => ( parameter.Type ?? (SyntaxNode)parameter ),

```

Comment:
Other end of the duplicate (https://github.com/Brightspace/D2L.CodeStyle/pull/841/files#r805400297)